### DISCUSSION TITLE
mesh_boolean performance issue on Windows OS

### DISCUSSION THREAD

**Discussion by cljbb** [libigl/libigl#2417] [Q&A]
I use the libigl mesh boolean function:
igl::copyleft::cgal::mesh_boolean(VA, FA, VB, FB, igl::MESH_BOOLEAN_TYPE_MINUS, VC, FC)
in my program, which is complied in Windows, Mac, and Linux.

The Mac and Linux versions run very fast but the Windows version runs very slowly. The processing between two meshes of 1MB each can take 1 hour or more, just an example.

I know that someone had the same issue before and it was caused by running in the debug mode. In my case, it is not due to the debug mode since I am in the release mode. If I used the debug mode, the 1 hour processing would become 10 hours or more.

I "install" libigl on my computer as suggested; there is no separate compilation. All the lib files are included as header files so all these required files are complied in the release mode together with my own code.

The only lib that I noticed that has to included in my project is libgmp-10.lib. It is downloaded automatically during the libigl setup process. There is also a libgmp-10.dll in the same folder but it does not seem to be used in my program.

Could anyone point out what might have caused the performance issue?
Do I need to manually complie libgmp and not to use the libgmp-10.lib?
As far as I can tell, libgmp-10.lib is the only one that I did not complie, so it might be a debug version although quite unlikely as it seems to given (downloaded) automatically.

Thank you in advance for your help. I tried many different options and I am out of ideas. My Windows program runs like a turtle, while my Mac and Linux programs can finish the same job in the blink of an eye.

**Comment by alecjacobson** [comment]
Not a windows user, so I don't have a lot of help to offer. It sure seems like something is in debug mode or something in CGAL is not configured correctly (you could try to make a minimal cgal-only example and see if it suffers the same issues).

**Reply by cljbb** [reply]
Thank you for the suggestion. What I did was actually very basic. I used these on Windows.
cd libigl/
mkdir build
cd build
cmake ../
make

Of course, I installed git for Windows to make sure packages can be downloaded automatically. That is it.

When I included relevant headers to my program and compile, the error message show that gmp library is missing. I looked in the libigl folders and found libgmp-10.lib. Included it to my project. Compile success.

My Mac and Linux libigl was based on a similar procedure but there is no need for the libgmp-10.lib (a Windows thing only).

All my source files are complied in the Release mode. I assume and believe all headers of libigl used in my program should all be complied in the Release mode also. The only thing I am not sure is the libgmp-10.lib because it is given in a folder during the "make" process. I believe it should be a release version.

Not sure if anyone using the Windows libigl can point out what I am doing wrong. Thank you.

**Comment by alecjacobson** [comment]
Just to be sure. Can you try

cmake ../ -DCMAKE_BUILD_TYPE=Release

**Reply by cljbb** [reply]
Thank you. Very good point. I didn't know there is this option. I will try and let you know.

By the way, I thought to use libigl we don't need to complie anything, correct?
Libigl is all header files; therefore, there is no need to compile the library in order to use it, correct?
If so, what difference is this "cmake ../ -DCMAKE_BUILD_TYPE=Release" going to make?

I am only using the header files, with the addition of libgmp-10.lib, which is not generated by cmake in my opinion. I suspect it is downloaded. Am I right?

**Comment by alecjacobson** [comment]
liibigl can be used in header-only mode (it's still eventually compiled,
just with your code), but dependencies like cgal, gmp etc. for those
modules may still need to be compiled as a static/dynamic library.

**Reply by cljbb** [reply]
Thank you for the information. I think I read it somewhere; however, what I could not understand was that the only library (lib file) needed in my program was libgmp-10.lib.

Anyway, I will do as you suggested and get back to you on the outcome. Thanks again. Much appreciated.

**Reply by cljbb** [reply]
I have re-setup libigl (fresh download and setup) with this option "cmake ../ -DCMAKE_BUILD_TYPE=Release". I did not run the last command "make" because it is not needed to compile my program.

The performance issue remains; there is no change; my Windows version program still runs very slow on mesh boolean even with small (low poly count) mesh input.

Is there anything else that I should try?

By the way, I don't think "cmake ../ -DCMAKE_BUILD_TYPE=Release" compiled any code; it is only for downloading needed packages.

**Comment by alecjacobson** [comment]
I'm a bit confused. I would recommend following the build instructions for
libigl to build the tutorials this would be

git clone git@github.com:libigl/libigl.git
cd libigl
mkdir build
cd build
cmake ../ -DCMAKE_BUILD_TYPE=Release
make
./bin/609_Boolean

**Reply by cljbb** [reply]
Thank you for the suggestion. May I check what is this command "git clone ***@***.***:libigl/libigl.git".

Is ***@***.*** an email address? Why is that needed to download the libigl?

I downloaded libigl from github manually.

**Reply by alecjacobson** [reply]
Oh that seems like an error. Fixed it.

**Reply by cljbb** [reply]
Thank you. I tried to clone but got this.

C:\Users\iiuuf>git clone git@github.com:libigl/libigl.git
Cloning into 'libigl'...
The authenticity of host 'github.com (20.205.243.166)' can't be established.
ED25519 key fingerprint is SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU.
This key is not known by any other names.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes (I entered yes here)
Warning: Permanently added 'github.com' (ED25519) to the list of known hosts.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

<<<< >>>>
It seems that I don't have permission to copy.

**Reply by cljbb** [reply]
@alecjacobson , here is an update.

I have not done this "git clone git@github.com:libigl/libigl.git" because there is a permission problem.

I debugged into mesh_boolean (the very slow function in my Windows program) and found that most of the processing time is in remesh_self_intersections().

Based on previous issues reports, I found that another user had the same problem in Sep 2021. See "igl::copyleft::cgal::mesh_boolean -> really slow on tiny sample ... is it normal ? #1897"

You tried to help him; however, the discussion seems to be pending. There is no confirmed solution.
